translation(JS): web/javascript/reference/statements/import (#498)
webdoky/content: files/uk/web/javascript/reference/statements/import/index.md

## Синтаксис

```js
import defaultExport from "module-name";
import * as name from "module-name";
import { export1 } from "module-name";
import { export1 as alias1 } from "module-name";
import { default as alias } from "module-name";
import { export1 , export2 } from "module-name";
import { export1 , export2 as alias2 , [...] } from "module-name";
import { "string name" as alias } from "module-name";
import defaultExport, { export1 [ , [...] ] } from "module-name";
import defaultExport, * as name from "module-name";
import "module-name";
```

- `defaultExport`
  - : Назва, що вказує на усталений експорт із модуля. Повинна бути дійсним ідентифікатором JavaScript.
- `module-name`
  - : Модуль, з якого відбувається імпорт. Обчислення вказівника залежить від середовища. Часто це відносний чи абсолютний URL до файлу `.js`, що містить модуль. В середовищі Node імпорти без розширення часто вказують на пакунки у `node_modules`. Деякі пакувальні програми можуть допускати імпортування файлів без розширення — краще перевіряти налаштування середовища. Дозволяються лише рядки, виділені одинарними чи подвійними лапками.
- `name`

>   - : Назва об'єкта модуля, який під час звертання до імпортів буде вжито як свого роду простір імен. Повинен бути дійсним ідентифікатором JavaScript.

Reviewer: ```suggestion
  - : Назва об'єкта модуля, яку під час звертання до імпортів буде вжито як свого роду простір імен. Повинна бути дійсним ідентифікатором JavaScript.
```

- `exportN`
  - : Назви експортованих значень до імпорту. Така назва може бути як ідентифікатором, так і рядковим літералом — залежить від того, що саме модуль `module-name` оголосив до експорту. Якщо це рядковий літерал — йому слід призначити псевдонім, який уже буде дійсним ідентифікатором.
- `aliasN`
  - : Псевдонім, що вказує на іменовані імпорти. Має бути дійсним ідентифікатором JavaScript.

## Опис

> Параметр `name` позначає ім'я "модульного об'єкта", що буде вжито як свого роду простір імен, для звертання до експортованих значень. Параметри `export` вказують окремі названі експортні об'єкти, в той час, як синтаксис `import * as name` імпортує їх всіх. Нижче буде наведено приклади уточнення цього синтаксису.

Reviewer: ```suggestion
Параметр `name` позначає ім'я "модульного об'єкта", що буде вжито як свого роду простір імен для звертання до експортованих значень. Параметри `export` вказують окремі названі експортні об'єкти, в той час, як синтаксис `import * as name` імпортує їх всіх. Нижче буде наведено приклади уточнення цього синтаксису.
```

Оголошення `import` допускаються лише нагорі модуля, і лише у модульних файлах. Якщо оголошення `import` зустрічається у немодульному контексті (як от у скриптових файлах, `eval`, `new Function`, які після розбору мають оформитися у "скрипт" чи "функцію") — викидається помилка `SyntaxError`. Аби завантажити модуль у немодульному контексті — слід використовувати [динамічний імпорт](/en-US/docs/Web/JavaScript/Reference/Operators/import).

> Оголошення `import` сконструйовані таким чином, щоб бути синтаксично жорсткими (зокрема: допускаються лише літеральні рядкові вказівники, і лише на верхньому рівні, і за умови, що всі прив'язки являються ідентифікаторами). Це дає можливість статично аналізувати модулі та синхронно їх компонувати, іще до їхнього обчислення. Це — ключова особливість, необхідна аби зробити модулі асинхронними по своїй природі, що забезпечує такий функціонал, як [`await` верхнього рівня](/en-US/docs/Web/JavaScript/Guide/Modules#top_level_await).

Reviewer: ```suggestion
Оголошення `import` сконструйовані таким чином, щоб бути синтаксично жорсткими (зокрема: допускаються лише літеральні рядкові вказівники, і лише на верхньому рівні, і за умови, що всі прив'язки є ідентифікаторами). Це дає можливість статично аналізувати модулі та синхронно їх компонувати, іще до їхнього виконання. Це — ключова особливість, необхідна, аби зробити модулі асинхронними за природою, що дає змогу працювати функціональності штибу [`await` верхнього рівня](/en-US/docs/Web/JavaScript/Guide/Modules#await-verkhnioho-rivnia).
```

Author: Докинув окремим комітом тут https://github.com/webdoky/content/pull/498/commits/8ecf3faefdb8f89ea1ee22083aae82aa100fffb6

Оголошення `import` сконструйовані таким чином, щоб бути синтаксично жорсткими (зокрема: допускаються лише літеральні рядкові вказівники, і лише на верхньому рівні — адже всі прив'язки є ідентифікаторами). Це дає можливість статично аналізувати модулі та синхронно їх компонувати, іще до їхнього виконання. Це — ключова особливість, необхідна аби зробити модулі асинхронними за природою, що дає змогу працювати функціональності штибу [`await` верхнього рівня](/uk/docs/Web/JavaScript/Guide/Modules#top_level_await).

Існує чотири форми оголошень `import`:

+- [Іменований імпорт](#named_import): `import { export1, export2 } from "module-name";`
+- [Усталений імпорт](#default_import): `import defaultExport from "module-name";`

> +- [Імпорт простору імен](#namespace_import): `import * as name from "module-name";`

Reviewer: ```suggestion
+- [Імпорт простору імен](#import-prostoru-imen): `import * as name from "module-name";`
```

> +- [Імпорт заради побічних ефектів](#import_a_module_for_its_side_effects_only): `import "module-name";`

Reviewer: ```suggestion
+- [Імпорт заради побічних ефектів](#importuvannia-modulia-lyshe-zarady-yoho-pobichnykh-efektiv): `import "module-name";`
```

Нижче наведено приклади для пояснення синтаксису.

### Іменований імпорт

> Нехай дано певне значення під назвою `myExport`, експортоване з модуля `my-module` або явно за допомогою інструкції {{JSxRef("Statements/export", "export")}}, або неявно (як `export * from 'another.js'`). Такий код вставить значення `myExport` у поточну область.

Reviewer: ```suggestion
Нехай дано певне значення під назвою `myExport`, експортоване з модуля `my-module` або явно за допомогою інструкції {{JSxRef("Statements/export", "export")}}, або неявно (як `export * from 'another.js'`). Такий код додасть значення `myExport` у поточну область.
```

```js
import { myExport } from '/modules/my-module.js';
```

З одного й того ж модуля можна імпортувати декілька імен.

```js
import { foo, bar } from '/modules/my-module.js';
```

> Можна перейменувати експорт під час імпортування його значення. Наприклад, цей код вставляє значення `shortName` у поточну область видимості.

Reviewer: ```suggestion
Можна перейменувати експорт під час імпортування його значення. Наприклад, цей код додає значення `shortName` у поточну область видимості.
```

```js
import {
  reallyReallyLongModuleExportName as shortName,
} from '/modules/my-module.js';
```

Модуль також може експортувати рядковий літерал, який не є дійсним ідентифікатором. В цьому випадку знадобиться звертатися до нього за псевдонімом, аби мати змогу користуватися ним у поточному модулі.

```js
// /modules/my-module.js
const a = 1;
export { a as "a-b" };
```

```js
import { "a-b" as a } from "/modules/my-module.js";
```

> **Примітка:** Інструкція `import { x, y } from "mod"` не є еквівалентною до `import defaultExport from "mod"` і потім деструктурування `x` та `y` із `defaultExport`. Усталений та іменований імпорт — це різні синтаксичні конструкції модулів JavaScript.

### Усталений імпорт

> Усталений експорт слід імпортувати за допомогою відповідного синтаксису усталеного імпорту. Найпростіший варіант безпосередньо імпортує усталене значення:

Reviewer: ```suggestion
Усталений експорт слід імпортувати за допомогою відповідного йому синтаксису усталеного імпорту. Найпростіший варіант безпосередньо імпортує усталене значення:
```

Також можливо вказувати усталений імпорт разом з імпортом простору імен чи іменованими імпортами. Проте в таких випадках усталений імпорт слід вказувати першим. Наприклад:

```js
import myDefault, * as myModule from '/modules/my-module.js';
// myModule.default та myDefault вказують на одну прив'язку
```

...або ж:

```js
import myDefault, { foo, bar } from '/modules/my-module.js';
```

Імпортування значення, яке називається `default`, має ефект, ідентичний усталеному імпортові. Йому необхідно призначати псевдонім, оскільки `default` — це зарезервоване слово.

 ```js
import { default as myDefault } from '/modules/my-module.js';
 ```

### Імпорт простору імен

Наступний код вставляє `myModule`, який містить всі значення, експортовані з модуля `/modules/my-module.js`, у поточну область.

```js
import * as myModule from '/modules/my-module.js';
```

В цьому випадку `myModule` є об'єктом _простору імен_, який вміщує всі експортовані значення як звичайні властивості. Наприклад, якщо імпортований вище модуль містить експорт `doAllTheAmazingThings()`, його можна було б викликати таким чином:

```js
myModule.doAllTheAmazingThings();
```

> Значення `myModule` — це [запечатаний](/uk/docs/Web/JavaScript/Reference/Global_Objects/Object/isSealed) об'єкт, чиїм прототипом є `null`. Всі його ключі є [перелічуваними](/uk/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) в лексикографічному порядку (тобто в послідовності, згідно з якою працює [`Array.prototype.sort()`](/uk/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#description)), причому усталений експорт буде доступний за ключем `default`.

Reviewer: ```suggestion
Значення `myModule` — це [запечатаний](/uk/docs/Web/JavaScript/Reference/Global_Objects/Object/isSealed) об'єкт, чиїм прототипом є `null`. Всі його ключі є [перелічуваними](/uk/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) в лексикографічному порядку (тобто в послідовності, згідно з якою працює [`Array.prototype.sort()`](/uk/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#opys)), причому усталений експорт буде доступний за ключем `default`.
```
